### PR TITLE
fix ability to mix/merge json env variables with env_nested_delimiter…

### PR DESCRIPTION
**Pull request by Air-Mark**
## Change Summary
fix ability to mix/merge json env variables with variables defined according to env_nested_delimiter

## Related issue number
Found a corner-case when we are mixing 2 types of nested environment variables definition.
#3159

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by Air-Mark** [comment]
please review

File: pydantic/env_settings.py [hunk 1]
```diff
@@ -232,7 +235,16 @@ def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[st
             env_var = result
             for key in keys:
                 env_var = env_var.setdefault(key, {})
-            env_var[last_key] = env_val
+
+            if env_val:
+                try:
+                    env_val = settings.__config__.json_loads(env_val)
+                except JSONDecodeError:
+                    ...
```
Comment (samuelcolvin) [inline pydantic/env_settings.py:243]:
```suggestion
                    pass
```

File: pydantic/env_settings.py [hunk 2]
```diff
@@ -232,7 +235,16 @@ def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[st
             env_var = result
             for key in keys:
                 env_var = env_var.setdefault(key, {})
-            env_var[last_key] = env_val
+
+            if env_val:
+                try:
+                    env_val = settings.__config__.json_loads(env_val)
+                except JSONDecodeError:
```
Comment (samuelcolvin) [inline pydantic/env_settings.py:242]:
```suggestion
                except ValueError:
```
More consistent to just use `ValueError`

File: pydantic/env_settings.py [hunk 3]
```diff
@@ -232,7 +235,16 @@ def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[st
             env_var = result
             for key in keys:
                 env_var = env_var.setdefault(key, {})
-            env_var[last_key] = env_val
+
+            if env_val:
+                try:
```
Comment (samuelcolvin) [inline pydantic/env_settings.py:240]: The problem here is we're calling `json.loads` on every field which:
1. could have some weird side effects - e.g. you want `"foobar"` but `json_loads` converts it to `foobar`
2. It could effect performance

We could:
* not do this and document it?
* check if `env_val` starts with `{` or `[` and only parse json in that case

**Review by samuelcolvin — COMMENTED** [review]
I get the request, but I'm not sure about the best way to proceed.

**Comment by samuelcolvin** [comment]
please update.

**Comment by samuelcolvin** [comment]
I think this overlaps a lot with #4216, but that solution is more complete, even that we're deferring to V2.
